python/adjoint/objective.py

```diff
@@ -21,152 +41,231 @@ def register_monitors(self, frequencies):
+        self._direction_vector = direction_vector if self.forward else -1 * direction_vector
+        return self._monitor
 
     def place_adjoint_source(self, dJ):
         dJ = np.atleast_1d(dJ)
-        dt = self.sim.fields.dt
-        direction_scalar = -1 if self.forward else 1
-        if self.kpoint_func is None:
-            if self.normal_direction == 0:
-                k0 = direction_scalar * mp.Vector3(x=1)
-            elif self.normal_direction == 1:
-                k0 = direction_scalar * mp.Vector3(y=1)
-            elif self.normal_direction == 2:
-                k0 = direction_scalar * mp.Vector3(z=1)
-        else:
-            k0 = direction_scalar * self.kpoint_func(self.time_src.frequency,
-                                                     1)
         if dJ.ndim == 2:
             dJ = np.sum(dJ, axis=1)
-        da_dE = 0.5 * self.cscale  # scalar popping out of derivative
+        time_src = self._create_time_profile()
+        da_dE = 0.5 * self._cscale
+        scale = self._adj_src_scale()
 
-        scale = adj_src_scale(self, dt)
+        if self.kpoint_func:
+            eig_kpoint = self.kpoint_func(time_src.frequency, self.mode)
+        else:
+            eig_kpoint = self._direction_vector
 
-        if self.frequencies.size == 1:
+        if self._frequencies.size == 1:
             amp = da_dE * dJ * scale
-            src = self.time_src
+            src = time_src
         else:
             scale = da_dE * dJ * scale
             src = FilteredSource(
-                self.time_src.frequency,
-                self.frequencies,
+                time_src.frequency,
+                self._frequencies,
                 scale,
-                dt,
+                self.sim.fields.dt,
             )
             amp = 1
-
-        self.source = [
-            mp.EigenModeSource(
-                src,
-                eig_band=self.mode,
-                direction=mp.NO_DIRECTION,
-                eig_kpoint=k0,
-                amplitude=amp,
-                eig_match_freq=True,
-                size=self.volume.size,
-                center=self.volume.center,
-                **self.EigenMode_kwargs,
-            )
-        ]
-        return self.source
+        source = mp.EigenModeSource(
+            src,
+            eig_band=self.mode,
+            direction=mp.NO_DIRECTION,
+            eig_kpoint=-1 * eig_kpoint,
+            amplitude=amp,
+            eig_match_freq=True,
+            size=self.volume.size,
+            center=self.volume.center,
+            **self.eigenmode_kwargs,
+        )
+        return [source]
 
     def __call__(self):
-        self.time_src = create_time_profile(self)
-        direction = mp.NO_DIRECTION if self.kpoint_func else mp.AUTOMATIC
+        if self.kpoint_func:
+            kpoint_func = self.kpoint_func
+            overlap_idx = self.kpoint_func_overlap_idx
+        else:
+            kpoint_func = lambda *not_used: self._direction_vector
```

Comment:
This doesn't seem right — the k point is dimensionful, so the default guess is scaled by `fcen*(refractive index)`.

For comparison, this is the default guess in `mpb.cpp`: https://github.com/NanoComp/meep/blob/83181144fdfdd76d295139da2f4aa0ea6b192aaf/src/mpb.cpp#L487